### PR TITLE
fix: better error loading of char data

### PR DESCRIPTION
**Pull request by vaab**
The user-defined function is now given a reject callback to trigger error
conditions while loading new character data. Please note that the
running code (not the user-defined function) is also given a chance to
reject the promise (and thus trigger the calling of the error callback)
when it receives a second loading request before the current one is
finished.

This should fix the issues with the error being reported only on next call 
if user-defined function doesn't handle errors itself correctly.

Please note that it doesn't provide a radical change of the API of the user defined function as I suggested in an open issue #41 ... and thus should remain mostly compatible with any custom made user function.

### CODE REVIEW

**Comment by vaab** [comment]
I just see that I didn't include the builds result with this patch applies (I'm talking of the new ``dist/*``). I can add them if you need them.

**Comment by chanind** [comment]
No worries, no need to add the `dist/*` files. I'll test this out some more later tonight. Thanks for investigating and providing a PR!

File: src/HanziWriter.js [hunk 1]
```diff
@@ -204,11 +209,14 @@ HanziWriter.prototype._loadCharacterData = function(char) {
...
     this.cancelLoadingCharData = reject;
-    const returnedData = this._options.charDataLoader(char, resolve);
+    const returnedData = this._options.charDataLoader(char, resolve, reject);
     if (returnedData) resolve(returnedData);
   }).then((data) => {
     this.isLoadingCharData = false;
     return data;
+  }, () => {
+    this.isLoadingCharData = false;
+    return Promise.reject.apply(Promise, arguments);
```
Comment (chanind) [inline src/HanziWriter.js]: Is this the same as `Promise.reject(arguments)`?

Comment (chanind) [reply src/HanziWriter.js]: nm, just realized this is `apply` instead of `call`, ignore previous comment :)

File: src/defaultCharDataLoader.js [hunk 2]
```diff
@@ -4,17 +4,24 @@
...
+  };
   xhr.onreadystatechange = () => {
     // TODO: error handling
-    if (xhr.readyState === 4 && xhr.status === 200) {
+    if (xhr.readyState !== 4) return;
+
+    if (xhr.status === 200) {
       onLoad(JSON.parse(xhr.responseText));
+    } else if (onError) {
+      onError(xhr);
```
Comment (chanind) [inline src/defaultCharDataLoader.js]: 👍

File: src/HanziWriter.js [hunk 3]
```diff
@@ -178,6 +179,10 @@ HanziWriter.prototype.setCharacter = function(char) {
 
     if (this._options.showCharacter) this._characterRenderer.showImmediate();
     if (this._options.showOutline) this._outlineRenderer.showImmediate();
+  }, () => {
+    if (this._options.onCharDataLoaderError) {
```
Comment (chanind) [inline src/HanziWriter.js]: There's a helper in `./utils` called `callIfExists()` that you can use here. It would look something like:
`callIfExists(this._options.onCharDataLoaderError, ...args)`

File: src/HanziWriter.js [hunk 4]
```diff
@@ -178,6 +179,10 @@ HanziWriter.prototype.setCharacter = function(char) {
 
     if (this._options.showCharacter) this._characterRenderer.showImmediate();
     if (this._options.showOutline) this._outlineRenderer.showImmediate();
+  }, () => {
```
Comment (chanind) [inline src/HanziWriter.js]: Could you use the `(...args) => {` style here? babel should essentially turn it into the same result in the end, and it's a bit more readable IMO since it makes it clear this function is expecting arguments.

File: src/HanziWriter.js (hunk 1 above)
Comment (chanind) [inline src/HanziWriter.js]: same here, this could be:
```js
(...args) => {
  this.isLoadingCharData = False;
  return Promise.reject(...args);
}
```

File: src/HanziWriter.js [hunk 5]
```diff
@@ -12,6 +12,7 @@ const { assign, isMSBrowser, timeout } = require('./utils');
 
 const defaultOptions = {
   charDataLoader: defaultCharDataLoader,
+  onCharDataLoaderError: null,
```
Comment (chanind) [inline src/HanziWriter.js]: nit: could this be called `onLoadCharDataError` instead? That way it looks more consistent with `isLoadingCharData`.

File: src/HanziWriter.js (hunk 4 above)
Comment (chanind) [reply src/HanziWriter.js]: This logic seems like it might make more sense in the error handler you added in `_loadCharacterData()` below. that way all the logic for loading character data and handling errors happens within `_loadCharacterData()`.

File: src/HanziWriter.js [hunk 6]
```diff
@@ -204,11 +209,14 @@ HanziWriter.prototype._loadCharacterData = function(char) {
   this.isLoadingCharData = true;
   return new Promise((resolve, reject) => {
     this.cancelLoadingCharData = reject;
-    const returnedData = this._options.charDataLoader(char, resolve);
+    const returnedData = this._options.charDataLoader(char, resolve, reject);
     if (returnedData) resolve(returnedData);
   }).then((data) => {
     this.isLoadingCharData = false;
```
Comment (chanind) [inline src/HanziWriter.js]: While we're here, how about adding a `onLoadCharDataSuccess` callback here too? That way it's easy to add a loading spinner if developers want to.

File: src/HanziWriter.js (hunk 4 above)
Comment (vaab) [reply src/HanziWriter.js]: no prob, I'm a newbie concerning babel and pretty much every thing serious in javascript (I mean outside standard javascript in browsers from a few years ago). Very happy to learn.

I'm happy also to move the logic here to ``_loadCharacterData``, but we have to prevent the then in this first function (``setCharacter()``) to set of : I can't see what I'm supposed to return in the rejected callback at the end of ``_loadCharacterData()`` that won't trigger ``.then()`` from ``setCharacter`` function or complain because of a missing ``reject`` callback in ``setCharacter()`` function.  With this perspective it seems quite natural to me to use the original logic I used as it simply works and I can't see a direct solution with your proposal. See my problem ? I might have missed a simple solution.

File: src/HanziWriter.js (hunk 5 above)
Comment (vaab) [reply src/HanziWriter.js]: Of course.

File: src/HanziWriter.js (hunk 3 above)
Comment (vaab) [reply src/HanziWriter.js]: nice !

File: src/HanziWriter.js (hunk 6 above)
Comment (vaab) [reply src/HanziWriter.js]: Hmm, I see... I'll do that.

File: src/defaultCharDataLoader.js (hunk 2 above)
Comment (vaab) [reply src/defaultCharDataLoader.js]: did you notice that the ``onError`` callback gets called in 3 different occasions.
- once by might be by ``_loadCharacter`` implementation whenever a new loading is asked and the previous is not finished.
- and twice in the ``defaultCharDataLoader``: in case of ``XmlHttpRequest.onerror`` event, and if the request is successfully answered but the HTTP error status in not '200'.
If you can distinguish from these 3 cases, it still one unique function getting called with 3 different way (different args and different ``this``). I didn't like that much, but I figured it gave a great deal of tinkerability without adding to much callbacks to deal with for very similar problems in the outer API of your library.

File: src/HanziWriter.js (hunk 4 above)
Comment (vaab) [reply src/HanziWriter.js]: And as a last connected thought : I saw that many function of HanziWriter are not returning promises where they easily could and this would provide many easy and natural ways to have these callbacks. For instance, why not return the promise created at the end of ``setCharacter()`` ? in this light, it seems very natural to pass along the resolved and the rejected state so as to leave the caller of ``setCharacter()`` to decide if he wants to hook a callback on both failure or success of the resolution of this promise. What do you think of that ?

File: src/defaultCharDataLoader.js (hunk 2 above)
Comment (chanind) [reply src/defaultCharDataLoader.js]: Hrm that's a good point. I think the 2 calls in `defaultCharDataLoader` are fine, since they both pass the `xhr` object back for the developer to figure out what happened. Probably for `_loadCharacter` when the previous request isn't finished, it seems like that shouldn't be an error. The new character should just take precedence and the old loading should be ignored, I think. Maybe we can pass some parameters in the reject to indicate if it's being rejected because of a real error, and just ignore the reject otherwise?

File: src/HanziWriter.js (hunk 4 above)
Comment (chanind) [reply src/HanziWriter.js]: I think it's OK to still trigger then `then()` in `setCharacter`, but maybe we can just have an if-statement at the top to check if there's any data passed in and just exit the function if not? Otherwise we could add a `.catch()` at the end, like you said. It just seems strange that errors in setting up the character would be considered char data loading errors the way it's written currently. I think it's OK to return a promise from `setCharacter()` too if that makes it easier

**Comment by chanind** [comment]
Ok, no rush, and thanks so much again for contributing this!

File: src/defaultCharDataLoader.js (hunk 2 above)
Comment (vaab) [reply src/defaultCharDataLoader.js]: Yup, I like this suggestion much better. I'll do that.

File: src/HanziWriter.js (hunk 4 above)
Comment (vaab) [reply src/HanziWriter.js]: I returned a promise form ``setCharacter()``. Added the condition at the beginning of the ``.then()`` to ignore empty results. And thus use this condition and I am returning null in the ``_loadCharacterData`` rejection callback.

**Comment by vaab** [comment]
oups, it seems what I pushed is not the final version I intended. Please wait a moment before reviewing.

**Comment by vaab** [comment]
I've provided the new patch (rebased on master). Please make sure that this is what you want again as we have discussed a few point. In particular, the new approach to cancelation of a current ongoing char loading.

**Comment by vaab** [comment]
I'm concerned about the fact that now (in this last version of the PR) in ``setCharacter`` function, the ``_loadCharacter()`` promise ``then`` callback is not reacting if data received is null. This means it doesn't set up all internal variable of ``HanziWriter``. As a consequnence, from the library user's view, using ``setCharacter()``, we can't really know if we can use the widget and request a ``HanziWriter.loopCharacterAnimation`` for instance as it'll fail with same error than #50 ... because internal variable are not set.
Do you understand the issue ? We could ignore that and make all API function immune to unset variable by just canceling their operation. Or/and try to make ``setCharacter()`` return a promise that it would reject in case if failed... to set the character. What do you think of that ?

**Comment by chanind** [comment]
What you're saying makes sense. All those methods should call `_withData()` which looks at `this._withDataPromise`. When we fail to load data, maybe we can set `this. _withDataPromise = null` and then in `_withData()` throw an exception that says something like `Failed to load data. Please call setCharacter() again to try reloading`? Either that, or we can auto-retry loading the character data or something. I'll take a more in-depth look in a few hours.

**Comment by chanind** [comment]
I added some logic to debounce subsequent requests on top of what you wrote, and moved the loading into a separate object since it was getting too complicated. This way, if there are multiple requests to load character data at the same time, the loader will only pay attention to the last one. I also modified `this._withData()` so that it will try to reload the character if loading failed and then the users tries to call a method on it. I tested this out manually, but I'll work on adding some test coverage to this in the next few days too. Thanks for your hard work on this!

**Comment by vaab** [comment]
Many thanks for the merge. Although, I have several concerns about your last commit as I don't understand how it is supposed to break away of an infinite loop (because of first lines of ``_withData()`` reloading). And in my current usage of your lib, without any changes, I get forced indeed in an infinite javascript loop in my test app... but I may have to change the way I'm using the lib.

Apart from the refactor, I'm not sure I understand the intention behind your last commit:  as I read you, you have added some logic "to debounce the subsequent requests"... which I may not understand fully. The code you merged would already give priority to the last call and cancel silently (as you suggested) any prior undergoing calls.

Sorry if I mis-understood something.

**Comment by chanind** [comment]
I didn't think it would cause an infinite loop because loading data happens asynchronously, but I see now that if an error is thrown synchronously it could cause an infinite loop because it will immediately try to reload again on the next line. I think adding another if-statement before https://github.com/chanind/hanzi-writer/commit/262f879492f0af929e8114b1c6e9dc9c52966333#diff-39b0e61ae24a34377369a66c7615d483R219 should be able to fix synchronous errors too. I'll add some test cases for these scenarios later before setting up another release. The reason for doing a retry is for the following scenario: Imagine a  user tries to set up a HanziWriter instance and loading fails. Then, a few seconds later they call `writer.animateCharacter()`, but the character has already failed to load. We can't animate the character since we weren't able to load it, so we need to throw an exception or something to tell the user they need to reload the character before trying to animate it again. However, we can just try reloading automatically and save the user the trouble of having to call `writer.setCharacter(...)` first again.

The reason for debouncing is to solve the issue you brought up where onError could be called if a request gets canceled before it's done, which seems wrong. Instead, if a character is loading, and then the user tries to load a new character before the previous loading request is finished, we should stop trying to load the old character and just focus on loading the new character. The debounce is there so that when the old character finally loads (or fails to load) we don't try to draw it. Even worse, if we don't debounce, it's possible the the old character could load after the new character finishes loading, and then we'll have drawn the wrong character. For instance, if the user calls: `writer.setCharacter('我')` and then 1ms later calls `writer.setCharacter('他')`, we should load 他 because that was the last character requested and forget about loading 我.

Another thought I had was that if the user doesn't provide an `onLoadCharError` callback, we should raise an exception when we catch an error. That way it will show up in their javascript error console so they'll be aware the error occurred. Currently if there's no `onLoadCharError` provided the user won't be able to know if loading failed and would be confused why their HanziWriter is blank. What do you think?

**Comment by chanind** [comment]
I added some tests here: https://github.com/chanind/hanzi-writer/pull/52. I wasn't able to reproduce the infinite loop, even with a synchronous callback. I'm not sure if that's because I was testing in Chrome and tests under node.js - it's possible different Promise implementations react differently to being rejected synchronously. I added an extra check here https://github.com/chanind/hanzi-writer/pull/52/files#diff-39b0e61ae24a34377369a66c7615d483R223 just in case. Let me know if you're still seeing an infinite loop.

**Comment by vaab** [comment]
> However, we can just try reloading automatically and save the user the trouble of having to call 
> writer.setCharacter(...) first again.

Hmm, but if it failed loading once, loading twice won't help it, will it ? Simply retrying won't fix the issue only in very special and rare cases. The most common failure is if the character is not in the database for instance. Other common failure would be connection error to the server after 60s timeout. Why bother retrying ?  In your scenario ``animateForever()`` should simply fail if no character was set, or at least do nothing. I'm not expecting ``animateForever()`` to retry something that already failed previously. Sorry if I misunderstood a point you wanted to make.

Currently, ``setCharacter()`` returns a promise, but not the one expected actually : if the character was not setup, we have no way to know that in the code. A reject of the promise from ``setCharacter()`` could be quite expected. What do you think ? As a user I could write:

```
     hw.setCharacter($el.value()).then(() => hw.animateForever(), () => console.log('Harrr crap !'));
```

For the debounce part, all this makes sense, and was already implemented in my proposal, or I may have missed some corner cases, but it was the intention.

Yes, an exception seems to me the right way to go if no ``onLoadCharErr`` is provided. I saw that you did implement that and wrote test about. Thx a lot.

**Comment by chanind** [comment]
The way I see it, if character loading failed and then the user later tries calling a method like `animateForever()` we can either try again and then throw an error and fail, or just immediately throw an error and fail. If loading failed because of a character not being in the database then retrying won't help, but neither will immediately throwing an error. However if loading failed because of a network issue it's possible that retrying will load successfully. If we're in this situation at all it means we've already notified the developer via throwing an exception or via the `onLoadCharDataError` callback and they didn't act on the error, so there's not a lot of hope they'll act on the error the second time. This is probably a pretty small edge-case anyway, but I don't see any reason to *not* retry.

>  if the character was not setup, we have no way to know that in the code

We just added `onLoadCharDataError` and `onLoadCharDataSuccess` callbacks so that we can know when characters load or fail to load in code.

> Currently, setCharacter() returns a promise, but not the one expected actually

It returns a promise that resolves when the character loading is complete. What should it return instead? If you don't pass an `onLoadCharDataError` callback then the loading will throw an error which should propagate into the catch clause in your example above, no?

**Comment by vaab** [comment]
May I try to convince you that you should not retry ? this is how I see it :
- I'm not expecting ``animateForever()`` to retry, this is not its obvious first responsibility if we read its name (nor any other ``_withData()`` animation). So you are adding a second (rather complex) behavior to possibly a lot of methods.
  - As a consequence I get errors unrelated to what I'm doing (HTTP timeouts, errors in logs, ...)
  - I don't have control on when access is made to ``loadCharacter()`` and I don't have control on how to react to this second failure.
  - I get my load error callback called twice, and need to be warned of that : not all implementation of the error callback might support it (if I display an alert to the user, I need to be careful not to output it twice). As a user of your library, I would ask myself : "why is my callback called twice ?". This is even more probable, that ``setCharacter()`` is not currently returning a rejected promise. So having a ``setCharacter()`` followed by a ``animateForever()`` without checking if ``setCharacter()`` internally succeeded to load the character is more than likely.
  - unit testing is harder for a bunch of method relying on ``_withData`` as they have an invisible link to the setCharacter method.
- if any retry should be done it is ``loadCharacter()`` that should do it. Depending on the way to retrieve the character, by network for instance, and notice that TCP handle retries for you. If any method of retrieval fails, it is for a reason. Making the assumption that it could need a retry, is not trusting ``loadCharacter()`` and entering in some way in a defensive coding... 
- If your argument to retry is you can't see any reason not to retry, then why not retry 3 times then ? The problem for me is not to give any reason "not to retry" but more to find a reason to retry. The fact that it might help on some corner case is a bad reason for me, because then I can find you hundreds of corner case where "it might help" if we coded defensively.... 

Well, you have my thoughts on this point. No issues if we don't get to any agreement on this. I only hope that I managed to correctly transmit my concerns in a sensible way.

Sorry for the second part, I was not clear, I meant that ``setCharacter()`` should probably return a promise that should be rejected in case it failed to actually set the character. But I guess you are still thinking about newcomers to promises, and expect them to use the new ``onLoadCharDataSuccess`` to follow up. Which make sense to me, but this callback gets called before the widget is fully setup. I note that this is not a big issue as a possible replacement of a resolved promise callback on ``setCharacter``, because even when calling an animation in ``onLoadCharDataSuccess``, the resolution of the animation will get attached to the ``_withDataPromise`` that will ensure that it gets called after the character was fully loaded. 

And you are also right about ``onLoadCharDataError`` ! I forgot about this new feature. This definitively helps.

So you can forget about my remarks on ``setCharacter()`` promise. Even, if as you probably guess it, I would prefer to have a standard resolve/reject pattern.

To be honest I would probably also remove the ``setCharacter()`` call in the ``HanziWriter`` constructor (avoiding any complex behavior in initialisation of object is often nice...), or at least make it optional, or best, provide a factory that would create the widget and load a character (``HanziWriter.Create()``). If you want more background why, well, here is food for thought and is probably debatable:

- http://misko.hevery.com/code-reviewers-guide/flaw-constructor-does-real-work/
- https://softwareengineering.stackexchange.com/questions/224063
- https://softwareengineering.stackexchange.com/questions/159193
- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c41-a-constructor-should-create-a-fully-initialized-object
- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-factory

... like that I could really manage when I'm loading my characters. 

I can provide PR for this also. If you think that makes sense but you are worried about backward compatibility, at least a optional way to avoid the ``setCharacter()`` would be nice, and providing the factory right away can't hurt.

**Comment by vaab** [comment]
I can move this last concern to a new issue also if it makes sense to you. Let me do this.

**Comment by chanind** [comment]
> I get my load error callback called twice, and need to be warned of that : not all implementation of the error callback might support it (if I display an alert to the user, I need to be careful not to output it twice)

That's a good point, I hadn't thought about it being confusing to have the error callbacks called multiple times. I can see how that would be unexpected. I think that's a good enough reason to not retry. If you want to make a PR removing the retry feel free! Otherwise I can try to get to it later this week.

**Comment by chanind** [comment]
Autoretry was removed in #64
